### ISSUE TITLE
Should Blazor docs cover web.config transforms for compression scenarios?

### ISSUE THREAD

**Issue by guardrex** [dotnet/AspNetCore.Docs#20445]
See https://github.com/dotnet/AspNetCore.Docs/issues/20405#issuecomment-723508184.

We could cross-link the [our transform doc](https://docs.microsoft.com/aspnet/core/host-and-deploy/iis/transform-webconfig) and include the file examples provided. However, we'd have to maintain them. This is an advanced scenario, and we have a rather large (*and growing!*) set of Blazor docs to maintain. It might be best to have it in a community blog/tutorial and cross-linked from Awesome Blazor. That way, we could point to it as needed and not take on the management of it.

cc: @robertmclaws

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: be59faf1-5a20-b7e9-19ff-06589023bb15
* Version Independent ID: 250e7458-3c1d-0117-261d-c03ac136f16a
* Content: [Host and deploy ASP.NET Core Blazor WebAssembly](https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?view=aspnetcore-3.1)
* Content Source: [aspnetcore/blazor/host-and-deploy/webassembly.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/host-and-deploy/webassembly.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
In order for automated deployments to work, you HAVE to have the production `web.config` file in the project. If you want it to work both locally AND deployed, you **MUST** do config transforms during compilation and deployment. Unfortunately, config transforms are not enabled OOB in Blazor projects, you have to set them up manually.

**- Step 1:** Add the following to your Blazor Standalone `.csproj` file:
```xml
  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.DotNet.Xdt.Tools" Version="2.0.0" />
  </ItemGroup>

  <Target Name="ApplyXdtConfigTransform" BeforeTargets="_TransformWebConfig">
    <PropertyGroup>
      <_SourceWebConfig>$(MSBuildThisFileDirectory)Web.config</_SourceWebConfig>
      <_XdtTransform>$(MSBuildThisFileDirectory)Web.$(Configuration).config</_XdtTransform>
      <_TargetWebConfig>$(PublishDir)Web.config</_TargetWebConfig>
    </PropertyGroup>
    <Exec Command="dotnet transform-xdt --xml &quot;$(_SourceWebConfig)&quot; --transform &quot;$(_XdtTransform)&quot; --output &quot;$(_TargetWebConfig)&quot;" Condition="Exists('$(_XdtTransform)')" />
  </Target>
```

**- Step 2:** Change your `web.config` to look like this:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<configuration>
  <system.webServer>
    <httpProtocol>
      <customHeaders>
        <add name="Blazor-Environment" value="Development" />
      </customHeaders>
    </httpProtocol>
    <staticContent>
      <remove fileExtension=".dat" />
      <remove fileExtension=".dll" />
      <remove fileExtension=".json" />
      <remove fileExtension=".wasm" />
      <remove fileExtension=".woff" />
      <remove fileExtension=".woff2" />
      <mimeMap fileExtension=".dat" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".dll" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".json" mimeType="application/json" />
      <mimeMap fileExtension=".wasm" mimeType="application/wasm" />
      <mimeMap fileExtension=".woff" mimeType="font/woff" />
      <mimeMap fileExtension=".woff2" mimeType="font/woff2" />     
      <mimeMap fileExtension=".blat" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".js.gz" mimeType="application/javascript" />
      <mimeMap fileExtension=".dat.gz" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".dll.gz" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".json.gz" mimeType="application/json" />
      <mimeMap fileExtension=".wasm.gz" mimeType="application/wasm" />
      <mimeMap fileExtension=".blat.gz" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".html.gz" mimeType="text/html" />
      <mimeMap fileExtension=".css.gz" mimeType="text/css" />
      <mimeMap fileExtension=".ico.gz" mimeType="image/x-icon" />
      <mimeMap fileExtension=".svg.gz" mimeType="image/svg+xml" />
      <mimeMap fileExtension=".js.br" mimeType="application/javascript" />
      <mimeMap fileExtension=".dat.br" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".dll.br" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".json.br" mimeType="application/json" />
      <mimeMap fileExtension=".wasm.br" mimeType="application/wasm" />
      <mimeMap fileExtension=".blat.br" mimeType="application/octet-stream" />
      <mimeMap fileExtension=".html.br" mimeType="text/html" />
      <mimeMap fileExtension=".css.br" mimeType="text/css" />
      <mimeMap fileExtension=".ico.br" mimeType="image/x-icon" />
      <mimeMap fileExtension=".svg.br" mimeType="image/svg+xml" />
    </staticContent>
    <httpCompression>
      <dynamicTypes>
        <add mimeType="application/octet-stream" enabled="true" />
        <add mimeType="application/wasm" enabled="true" />
      </dynamicTypes>
    </httpCompression>
    <rewrite>
      <outboundRules rewriteBeforeCache="true">
        <rule name="Add Vary Accept-Encoding" preCondition="PreCompressedFile" enabled="true">
          <match serverVariable="RESPONSE_Vary" pattern=".*" />
          <action type="Rewrite" value="Accept-Encoding" />
        </rule>
        <rule name="Add Encoding Brotli" preCondition="PreCompressedBrotli" enabled="true" stopProcessing="true">
          <match serverVariable="RESPONSE_Content_Encoding" pattern=".*" />
          <action type="Rewrite" value="br" />
        </rule>
        <rule name="Add Encoding Gzip" preCondition="PreCompressedGzip" enabled="true" stopProcessing="true">
          <match serverVariable="RESPONSE_Content_Encoding" pattern=".*" />
          <action type="Rewrite" value="gzip" />
        </rule>
        <preConditions>
          <preCondition name="PreCompressedFile">
            <add input="{HTTP_URL}" pattern="\.(gz|br)$" />
          </preCondition>
          <preCondition name="PreCompressedBrotli">
            <add input="{HTTP_URL}" pattern="\.br$" />
          </preCondition>
          <preCondition name="PreCompressedGzip">
            <add input="{HTTP_URL}" pattern="\.gz$" />
          </preCondition>
        </preConditions>
      </outboundRules>
      <rules>
        <rule name="Rewrite brotli file" stopProcessing="true">
          <match url="(.*)"/>
          <conditions>
            <add input="{HTTP_ACCEPT_ENCODING}" pattern="br" />
            <add input="{REQUEST_FILENAME}" pattern="\.(js|dat|dll|json|wasm|blat|htm|html|css|ico|svg)$" />
            <add input="{REQUEST_FILENAME}.br" matchType="IsFile" />
          </conditions>
          <action type="Rewrite" url="{R:1}.br" />
        </rule>
        <rule name="Rewrite gzip file" stopProcessing="true">
          <match url="(.*)"/>
          <conditions>
            <add input="{HTTP_ACCEPT_ENCODING}" pattern="gzip" />
            <add input="{REQUEST_FILENAME}" pattern="\.(js|dat|dll|json|wasm|blat|htm|html|css|ico|svg)$" />
            <add input="{REQUEST_FILENAME}.gz" matchType="IsFile" />
          </conditions>
          <action type="Rewrite" url="{R:1}.gz" />
        </rule>
      </rules>
    </rewrite>
  </system.webServer>
</configuration>
```
This version combines the static content remappings from the default `web.config` that the build process adds if one is not present, with the Rewrite rules that will not break the app locally.

**- Step 3:** Add a `web.{ENVIRONMENT}.config` and add the following:
```xml
<?xml version="1.0" encoding="utf-8"?>

<!-- For more information on using web.config transformation visit https://go.microsoft.com/fwlink/?LinkId=125889 -->

<configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">

  <system.webServer>
    <httpProtocol>
      <customHeaders>
        <add xdt:Transform="SetAttributes" xdt:Locator="Match(name)" name="Blazor-Environment" value="{YOUR ENVIRONMENT HERE}" />
      </customHeaders>
    </httpProtocol>
    <httpCompression xdt:Transform="Replace">
      <dynamicTypes>
        <remove mimeType="text/*" />
        <remove mimeType="application/javascript" />
        <remove mimeType="image/svg+xml" />
      </dynamicTypes>
      <staticTypes>
        <remove mimeType="text/*" />
        <remove mimeType="application/javascript" />
        <remove mimeType="image/svg+xml" />
      </staticTypes>
    </httpCompression>
    <rewrite>
      <rules>
        <rule xdt:Transform="InsertBefore(/configuration/system.webServer/rewrite/rules/rule[@name='Rewrite brotli file'])"
          name="Serve subdir">
          <match url=".*" />
          <action type="Rewrite" url="wwwroot\{R:0}" />
        </rule>
        <rule xdt:Transform="InsertAfter(/configuration/system.webServer/rewrite/rules/rule[@name='Rewrite gzip file'])"
          name="SPA fallback routing" stopProcessing="true">
          <match url=".*" />
          <conditions logicalGrouping="MatchAll">
            <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="true" />
          </conditions>
          <action type="Rewrite" url="wwwroot\" />
        </rule>
      </rules>
    </rewrite>
  </system.webServer>

</configuration>
```
This version adds the environment headers, replaces the compression section, and inserts nodes at the beginning and end of the rewrite rules, so that the deployed version matches what is currently in the documentation.

That allows the app to work locally for debugging, and have the web.config transformed appropriately for deployment environments... while also working with the appsettings environment transform requirements.

Hope that makes it easy for you folks to capture in the documentation. HTH!

**Comment by guardrex** [comment]
I opened https://github.com/dotnet/AspNetCore.Docs/issues/20445.

Compression with **_an untransformed web.config file_** across environments is apparently not a supported scenario OOB for debugging. That's why I asked @captainsafia to comment.

The production `web.config` can be swapped into the deployment assets prior to deployment, changed on the hosting system after deployment but before starting the app pool, or exist/persist on the hosting system continuously (i.e., not deployed with the app). Transforms are one of several approaches that devs can find success with, including in automated deployment scenarios.

WRT to covering the transform approach, that's an advanced scenario given our current, ordinary level of coverage. I don't know if management would like it covered or not, so I opened that issue and marked it for triage. Given that we're right on the edge of 5.0 GA, I don't think that we'll get a quick answer. Then with the ⛄ *holidays* 🎁 fast approaching, it might be triaged early next year. Hard to say ... everyone is 🏃😅 like crazy right now. We'll keep an :ear: on the issue to see if they give it any ❤️ ... thanks for the suggestion.

**Comment by guardrex** [comment]
> Compression with an untransformed web.config file across environments is apparently not a supported scenario OOB for debugging. That's why I asked @captainsafia to comment.

Trying to catch up with the thread. Is there a repro for a project with the web.config that fails to debug?

**Comment by guardrex** [comment]
👋🏻 @captainsafia! Just use [either of the web.configs specified at the top of this post](https://github.com/dotnet/AspNetCore.Docs/issues/20405#issuecomment-721988548) in a Blazor project and debug locally.

Appreciate all of the quick responses and the engagement, you folks are awesome!

@Guardrex unfortunately none of the approaches you specified in your last post are tenable from an operations standpoint in anything but hobbyist situations. When a `web.config` does not exist in the project, the `dotnet publish` process generates a new one, which overwrites whatever you've previously uploaded. (I just confirmed this in my own Azure DevOps environment.) Any other support mechanism (Azure DevOps tasks, post-deployment scripts, manual intervention) are actually *more* complicated than the config transform process I outlined, which conceptually most developers should be familiar with.

IMO, I think we've uncovered that compression itself is an advanced scenario that requires its own in-depth article. Given all of the other constraints on the next few weeks, and that most developers would likely rather be able to debug their project than have optimized compression... it might make sense to pull compression from the docs for now. As you mentioned in the other issue, I'm happy to write a blog post about it to link from Awesome Blazor.

**Comment by guardrex** [comment]
Publish with ...

```xml
<PublishIISAssets>false</PublishIISAssets>
```

... and hopefully it won't produce one or leave the one in the `publish` folder alone. 🤞

IIRC, set to `true` to take a custom file from the project root (maybe 🤞 ... it's a been a while since I played with that property).

*hobbyist* :smile:

@captainsafia will let us know if the situation is so dire that we'd need to pull coverage ... or NOTE ... or WARNING ... we have options to leave the current coverage and surface something to readers on this.

WRT publishing something on the transform scenario, I think a blog post+Awesome Blazor would be best for *FAST* 🏃  coverage. The issue that I opened might not get triaged for a while. Even if it becomes a work item, I have a feeling that it will get a low priority. If so, that would definitely push it to next year.

**Comment by guardrex** [comment]
*From @cornem ...*

FWIW, I always debug Blazor (well, any ASP.NET Core app) running them
stand-alone, rather than through IIS (Express). In that case, the
web.config file is ignored.

**Comment by captainsafia** [comment]
I took a look at this. It is indeed correct that debugging fails when you have a custom web.config. However, it doesn't seem to be related to compression at all. Using the [built-in web.config](https://github.com/dotnet/aspnetcore/blob/1cc368414c3041fe5a158e995b9df2cefbe63f67/src/Components/WebAssembly/Sdk/src/targets/BlazorWasm.web.config) results in the VS not sending the opening message to the debugger which causes it to redirect to the homepage.

I think the issue is with IIS + custom web.config in general as opposed to the compression scenario in particular. For now, I'd recommend that folks debug with the standalone server even if they are deploying to IIS in prod. I'll investigate further to see what might be causing the issue here. I'm starting to suspect some of those URL rewrite rules....

**Comment by guardrex** [comment]
I'll mark on HOLD for the time being. If any reader writes in, I'll direct them to this issue.

One ❓ tho on that @captainsafia ... When you say "the standalone server," you mean Kestrel?

**Comment by robertmclaws** [comment]
It's definitely the first and last rewrite rules that are the problem. If you look at the config transforms I put in the last section of Step 3, there are two rules that cause it to fail that it adds back in the right order.

HTH!

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
This is not something we plan to do in the near future. You can learn more about our triage process and how we handle issues by reading our [Triage Process](https://github.com/dotnet/aspnetcore/blob/master/docs/TriageProcess.md) writeup.
